### PR TITLE
Cancel tower analysis

### PR DESCRIPTION
**Pull request by fevac**
## Description

Closes: https://github.com/Clinical-Genomics/trailblazer/issues/226

### Changed

- `cancel` supports NF-Tower cases

### How to prepare for test
- [x] ssh to hasta (depending on type of change)
- [x] activate stage: `us`
- [x] request trailblazer-stage on hasta: `paxa`
- [x] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b  cancel_tower -a
    ```

### How to test
- [x] Cancel not ongoing analysis: `trailblazer cancel 279435`. It should raise the following error: `TrailblazerError: Analysis 279435 is not running`
- [x] Cancel ongoing analysis with failed tower response: `trailblazer cancel 279435`. It should raise the following error: `Request failed for url {url}: Error: {error}`
- [x] Cancel ongoing analysis: `trailblazer cancel 271983`. It complete successfully with the following message: `Analysis 279435: cancelled successfully!`

## Review
- [X] tests executed by EFC
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: trailblazer/store/crud/update.py [hunk 1]
```diff
@@ -116,6 +116,20 @@ def cancel_ongoing_analysis(
...
+        """Cancel slurm analysis by cancelling all associated slurm jobs."""
+        for job in analysis.jobs:
+            if job.status in SlurmJobStatus.ongoing_statuses():
+                LOG.info(f"Cancelling job {job.slurm_id} - {job.name}")
+                cancel_slurm_job(analysis_host=analysis_host, slurm_id=job.slurm_id)
+
+    def cancel_tower_analysis(self, analysis: Analysis) -> None:
+        """Cancel a NF-Tower analysis."""
+        LOG.info(f"Cancelling Tower workflow for {analysis.family}")
+        self.query_tower(config_file=analysis.config_path, case_id=analysis.family).cancel
```
Comment (henrikstranneheim) [inline trailblazer/store/crud/update.py:131]: Will Tower cancel associated Slurm jobs?

File: trailblazer/store/crud/update.py [hunk 2]
```diff
@@ -116,6 +116,20 @@ def cancel_ongoing_analysis(
         )
         self.commit()
 
+    def cancel_slurm_analysis(
+        self, analysis: Analysis, analysis_host: Optional[str] = None
+    ) -> None:
+        """Cancel slurm analysis by cancelling all associated slurm jobs."""
```
Comment (henrikstranneheim) [inline trailblazer/store/crud/update.py:122]:
```suggestion
        """Cancel SLURM analysis by cancelling all associated SLURM jobs."""
```

File: trailblazer/store/crud/update.py (hunk 1 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/update.py:131]: If so add it to doc str

Comment (fevac) [reply trailblazer/store/crud/update.py:131]: I think it does. But I will test and make sure before resolving this

Comment (fevac) [reply trailblazer/store/crud/update.py:131]: It does. I'll update the docstring

Comment (fevac) [reply trailblazer/store/crud/update.py:131]: done

File: trailblazer/apps/tower/api.py [hunk 3]
```diff
@@ -102,6 +103,13 @@ def workflow(self) -> TowerWorkflowResponse:
             url = self.build_url(endpoint=self.workflow_endpoint)
             return TowerWorkflowResponse(**self.send_request(url=url))
 
+    @property
+    def cancel(self) -> None:
+        """Send a POST request to cancel a workflow."""
+        if self.meets_requirements:
```
Comment (seallard) [inline trailblazer/apps/tower/api.py:109]: Could this bool be given a more descriptive name? Was confused reading it, but feel free to disregard

**Review by seallard — APPROVED** [review]
Nice 👍 Was a little confused by some stuff, nothing about the logic though

File: trailblazer/apps/tower/api.py [hunk 4]
```diff
@@ -102,6 +103,13 @@ def workflow(self) -> TowerWorkflowResponse:
             url = self.build_url(endpoint=self.workflow_endpoint)
             return TowerWorkflowResponse(**self.send_request(url=url))
 
+    @property
```
[the remark this answers is not in the document]
Comment (fevac) [reply trailblazer/apps/tower/api.py]: good to know. Changed to method

File: trailblazer/apps/tower/api.py (hunk 3 above)
Comment (fevac) [reply trailblazer/apps/tower/api.py:123]: updated to `send_cancel_request`

**Comment by fevac** [comment]
tests passing

**Comment by fevac** [comment]
- Deployed to stage

image

- Deployed to production

image

**Pull request merged**